Improve release workflow and readme (#41)
bevyengine/bevy_github_ci_template: README.md

Once you complete the process, a new release will be available on GitHub, with the archives for each platform accessible as downloadable assets.

![github release](https://user-images.githubusercontent.com/104745335/268805270-ff824032-4191-4528-9d45-a5511fef4f94.png)

### Publish on Itch.io

To publish releases to itch.io, follow this release flow:

1. Create an API key at <https://itch.io/user/settings/api-keys>
2. In your GitHub repository, go to the "Settings" tab, click on "Secrets" under the "Actions" section in the sidebar, and add a repository secret named `BUTLER_CREDENTIALS`, with the API key as its value.
3. Create your game page on itch.io if you haven't already.
4. In the [release.yaml](./.github/workflows/release.yaml#L22) file, set the environment variable `add_itchio_release: true`.
5. Uncomment the `env.itchio_target` in [release.yaml](./.github/workflows/release.yaml#L25) and set it to your itch.io username and the name of the game on itch.io, separated by a slash (`/`). For example: `cart/build-a-better-buddy`. Double-check the URL of your game to ensure the name is correct.

> Once these steps are completed, any tag pushed to GitHub will trigger an itch.io release, and it will use the tag as the[user version](https://itch.io/docs/butler/pushing.html#specifying-your-own-version-number).

Reviewer: Missing space between `the` and `[user version]`.

Reviewer 2: ```suggestion
Once these steps are completed, any tag pushed to GitHub will trigger an itch.io release, and it will use the tag as the [user version](https://itch.io/docs/butler/pushing.html#specifying-your-own-version-number).
```

> To make the game visible on your itch.io page, go to your game's configuration on itch.io, and change the `Kind of project` to HTML also locate your uploaded web files, and check the box that says, `This file will be played in the browser`.

Reviewer: Shouldn't it be noted that this is only necessary if you publish a WASM version and want to make it playable in the itch.io game page?

I suspect that "To make the game visible on your itch.io page" means exactly that, but I've only made that connection because I already know what this step does.

Author: Done, I added a clarification

![Kind of project](https://user-images.githubusercontent.com/104745335/268805329-fb70e23e-44ee-4f2f-9d20-11d58ddeec9a.png)

![Play in browser](https://github-production-user-asset-6210df.s3.amazonaws.com/104745335/268780679-fa14874c-040b-41ff-8a04-71cf141970dc.png)

### Publish on Github pages

> To publish on GitHub Pages, follow these steps:

Reviewer: I think this part should be more clear that this is about a play-in-browser WASM version. Otherwise one may interpret that this is about creating a GitHub Pages site for downloading the game.

Author: True, fixed

> 1. In the [release.yaml](./.github/workflows/release.yaml#L28) file, set the environment variable `add_github_pages_release: true` and the `add_web: true`.

Reviewer: Looking at the CI code - isn't the variable name `publish_github_pages`? Also, I don't see `add_web` - maybe you mean `build_web`?

Also note that both these environment variables are `true` by default.

Author: The env were changed and I forgot to update the readme, fixed.

# Bevy GitHub CI Template

> Template for setting up continuous integration (CI) on a GitHub project for Bevy. This template enables you to test your code, build for the Web, Linux, Windows, MacOS, and publish to GitHub Releases, Itch.io, and GitHub Pages.

Author: true, fixed

It creates two workflows:

* [CI](#ci)
* [Release](#release)

## CI

Definition: [.github/workflows/ci.yaml](./.github/workflows/ci.yaml)

![ci workflow](https://user-images.githubusercontent.com/104745335/268799840-06b772e8-7901-4b86-9a88-e4afee8a0167.png)

This workflow runs on every commit to `main` branch, and on every PR targeting the `main` branch.

It will use Rust stable on Linux with caching between different executions. The following commands are executed:

* `cargo test`
* `cargo clippy -- -D warnings`
* `cargo fmt --all -- --check`

If you are using anything OS-specific or Rust nightly, you should update the [ci.yaml](./.github/workflows/ci.yaml) file  accordingly.

## Release

Definition: [.github/workflows/release.yaml](./.github/workflows/release.yaml)

### Run Workflow

This workflow runs every time you push a tag to your repo.

![Release workflow](https://user-images.githubusercontent.com/104745335/268805183-28e83868-fb19-4e5e-ae96-7553fa3ede3c.png)

Example using git:

```sh
git tag -a "v1.0.0" -m "First official release"
git push --tags
```

> You can use the workflow as it is, but check that the name of your repository matches the name of your binary. If they don't match, you can change the env variable [`binary`](.github/workflows/release.yaml#L7) in the release.yaml file. By default it will build for the Web, Linux, Windows, and MacOS, and then it will publish the archives on GitHub Release, Itch.io and the version web to Github Pages.

Reviewer 3: ```suggestion
If the name of your repository does not match the name of your binary, you should update the environment variable [`binary`](.github/workflows/release.yaml#L7) in the `release.yaml` file to your binary name. By default, the workflow will build for the Web, Linux, Windows, and MacOS. The builds will be uploaded to a GitHub Release.
```

> You can configure the [builds](.github/workflows/release.yaml#L15) and [publish](.github/workflows/release.yaml#L21) targets by changing the [env variables](.github/workflows/release.yaml#L4)  in the release.yaml file.

Reviewer 3: ```suggestion
You can configure the [build](.github/workflows/release.yaml#L15) and [publish](.github/workflows/release.yaml#L21) targets by changing the [environment variables](.github/workflows/release.yaml#L4)  in the `release.yaml` file.
```

#### Manual Workflow run in GitHub

> Also you can configure and triggered the workflow directly in your GitHub repo. Navigate to the `Actions` section, click `Release` on the sidebar, then press the `Run workflow` button and select the configuration for the build and publish targets you want.

Reviewer 3: ```suggestion
You can configure and trigger the workflow directly in your GitHub repository. Navigate to the `Actions` section, click `Release` on the sidebar, then press the `Run workflow` button and select the configuration for the build and publish targets you want.
```

![Run manual workflow](https://github-production-user-asset-6210df.s3.amazonaws.com/104745335/269753464-cc39382c-67de-43e9-90aa-6f66ca444732.png)

> The configuration in GitHub takes priority over the env variables in the [release.yaml](.github/workflows/release.yaml#L4) file, so you don't have to modify your env variables. The manual workflow also enables you to override the tag version eliminating the need to create another tag to trigger the workflow.

Reviewer 3: ```suggestion
The configuration in GitHub takes priority over the environment variables in the [release.yaml](.github/workflows/release.yaml#L4) file. A manual workflow run also enables you to override the tag version eliminating the need to create a tag to trigger the workflow.
```

> For the itch.io target, you can leave it empty if you have already set up the env variable, If not, you will need to write your `itchio_target` every time you trigger the workflow manually.

Reviewer 3: ```suggestion
If you want to publish to itch.io, you will either have to configure the environment variable `itchio_target`, or define the `itchio_target` input for every run.
```

### Build

You can build for the following platforms:

* For Linux and Windows, create a .zip archive containing the executable and the `assets` folder.

> * For MacOS, generate two versions: one for Intel x86_64 and one for Apple Silicon ARM64, each packaged in a .dmg image containing a .app file with the assets.

Reviewer 3: Maybe it would be better to create a universal binary for both targets? See https://github.com/NiklasEi/bevy_game_template/blob/main/.github/workflows/release.yaml#L34

Author: I don't know, I was testing with the universal binary, and I think having separate targets has more benefits. You get half the compilation times because it's done in parallel, and the user can choose their platform directly, resulting in a smaller binary.

* For the Web, produce a .zip archive with the WebAssembly (wasm) binary, JavaScript bindings, an HTML file for loading it, and the assets.

By default, it builds for all platforms, but you can disable specific targets by removing their name in the env variable [`build_for`](.github/workflows/release.yaml#L15) in the release.yaml file.

> For example, `build_for: web, linux, macos` will build for the Web, Linux and MacOS but don't will build for windows.

Reviewer 3: ```suggestion
For example, `build_for: web, linux, macos` will build for the Web, Linux and MacOS but will not build for windows.
```

2. Trigger the [release.yaml](./.github/workflows/release.yaml) workflow by pushing a tag.
3. In your GitHub repository, go to the `Settings` tab, then click on `Pages` in the sidebar. Navigate to the `Build and Deployment` section and select `Github Actions` as the `Source`. ![Github Pages](https://user-images.githubusercontent.com/104745335/270078892-a6da763e-5a4a-4f55-8d3c-41a14e8423e2.png)
4. Wait a few minutes and your page will be available at a URL following this structure: `https://<Your GitHub username>.github.io/<Name of your repository>/`

If you want to publish by manually triggering a workflow in GitHub, instead of doing the 1 and 2 steps, write all your configuration in the GitHub Workflow form and then run the workflow. Refer to the ['Manual Workflow Run in GitHub'](#manual-workflow-run-in-github).

## Adapting the Workflows for Your Project

If you'd like to use the GitHub workflows provided here for your own project, you may need to make a few adjustments:

1. For web builds, ensure that your project includes an `index.html` file under the `/wasm` directory.

> 2. Make sure that the name of your repository matches the name of your binary. If the name of your repository is different, simply change the env variable [`binary`](.github/workflows/release.yaml#L7) in the release.yaml file to match the name of your binary.

Reviewer 4: And for stuff like this, would GitHub `vars` help to also avoid changing the script? This way you could default to the repo name if a variable is not found in the repository (something like `${{ vars.BINARY_NAME || github.repo }}`). This same suggestion might apply for some other places where default values could be overwritten without changing scripts.

3. If your project doesn't have an `assets` folder:
    1. You can create one and add a `.gitkeep` file to it to enable you to push it to your repository.

>     2. Alternatively, if your project does not use assets, you can remove the `cp -r assets` statements from the build jobs in the workflow.

Reviewer 4: Would a "disable assets folder copy" input help with this? By default it would be false but users could turn it on and that step will not run. This way you don't need to alter the template for this use case. It could be too much for too little benefit though so feel free to ignore this.
